### ISSUE TITLE
source_location vs __builtin_FUNCTION()

### ISSUE THREAD

**Issue by huangqinjin** [microsoft/STL#3138]
**Describe the bug**

`source_location`  and `__builtin_FUNCTION()` behave differently when used in a default argument.

**Command-line test case**
```
C:\test>type main.cpp
#include <iostream>
#include <source_location>

using namespace std;

auto f1(const char* s = __builtin_FUNCTION()) { return s; }
void g1(const char* s = f1()) { printf("g1 [%s]\n", s); }

auto f2(source_location s = source_location::current()) { return s; }
void g2(source_location s = f2()) { printf("g2 [%s]\n", s.function_name()); }

int main()
{
    g1(); // output: g1 [main]
    g2(); // output: g2 []
}

C:\test>cl /std:c++20 main.cpp
Microsoft (R) C/C++ Optimizing Compiler Version 19.33.31630 for x64
Copyright (C) Microsoft Corporation.  All rights reserved.

main.cpp
Microsoft (R) Incremental Linker Version 14.33.31630.0
Copyright (C) Microsoft Corporation.  All rights reserved.

/out:main.exe
main.obj

C:\test>main
g1 [main]
g2 []
```

**Additional context**
Seems that it is due to `consteval`. 
Here is related discussion of Clang https://reviews.llvm.org/D129488.

**Comment by StephanTLavavej** [comment]
This might be a compiler bug - also, this might be affected when we're able to use the recently-implemented `__builtin_FUNCSIG`.

**Comment by frederick-vs-ja** [comment]
I'm sure that this is certainly a compiler bug, and nothing can be fixed within the library implementation ([Godbolt link](https://godbolt.org/z/TW1jhsfMx)). I've reported DevCom-10174292.

Perhaps the bug of MSVC is _not_ related to `consteval`, because it can be reproduced with `constexpr` and non-`constexpr` functions.

Clang is more buggy at this moment, while Gcc has fixed a similar bug recently.

**Comment by frederick-vs-ja** [comment]
I'm sure that this is certainly a compiler bug, and nothing can be fixed within the library implementation ([Godbolt link](https://godbolt.org/z/TW1jhsfMx)). I've reported DevCom-10224094.

Perhaps the bug of MSVC is _not_ related to `consteval`, because it can be reproduced with `constexpr` and non-`constexpr` functions.

Clang is more buggy at this moment, while Gcc has fixed a similar bug recently.

**Comment by StephanTLavavej** [comment]
I suspect that this was just fixed by internal MSVC-PR-456552 for VS 2022 17.7 Preview 1. I've reached out to @JonCavesMSFT for confirmation.

**Comment by fsb4000** [comment]
I believe the issue is fixed in VS 17.7 preview 1
```
g1 [main]
g2 [int __cdecl main(void)]
```
![изображение](https://github.com/microsoft/STL/assets/4289847/d4d28d67-c51c-40ff-93f6-7e6a80ecd13c)

**Comment by CaseyCarter** [comment]
> I believe the issue is fixed in VS 17.7 preview 1

Confirmed fixed. 🎉

**Comment by JonCavesMSFT** [comment]
[like]  Jonathan Caves reacted to your message:
________________________________
